### PR TITLE
Fix shared memory permission issue in a shared pod environment

### PR DESCRIPTION
**Pull request by XiaohanZhangCMU**
## Description of changes:

We run into read permission issues when the shared memory is created in a shared computer environment. We relax the retry condition to allow the creation of a new shared memory file under permission error. When creating a shared memory file, we make sure both LOCALS and FILELOCKS do not overlap with existing ones and they have the same prefix ints in the name. 

What makes it more complicated is when a program exited non-normally, some of the shared meomory files are cleaned up but some are not. To address that, we need to make sure all files in SHM_TO_CLEAN do not have duplicates.

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [ ] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

A complete list of testings done for this PR is put in this [doc](https://docs.google.com/document/d/1stz7MMN31nqFODqFf6FDAx1NuGVJ-nhv76FJhGtPIJ8/edit?tab=t.0).

### CODE REVIEW

File: streaming/base/shared/prefix.py [hunk 1]
```diff
@@ -112,6 +112,10 @@ def _check_and_find(streams_local: list[str], streams_remote: list[Union[str, No
             shm = SharedMemory(name, False)
         except FileNotFoundError:
             break
+        except PermissionError:
+            # When trying to attach to a shm file on shared cluster
```
Comment (snarayan21) [inline streaming/base/shared/prefix.py:116]:
```suggestion
            # Trying to attach to a shm file on shared cluster
```

File: streaming/base/shared/prefix.py [hunk 2]
```diff
@@ -112,6 +112,10 @@ def _check_and_find(streams_local: list[str], streams_remote: list[Union[str, No
...
+            # may run into permission issue, we need to keep search.
```
Comment (snarayan21) [inline streaming/base/shared/prefix.py:117]:
```suggestion
            # may run into permission issues. Continue searching.
```

**Review by snarayan21 — COMMENTED** [review]
this seems fine to me, but are you able to reproduce this with an example where this used to fail, and then installing streaming off this branch now succeeds?

**Comment by snarayan21** [comment]
closing after offline discussion with @XiaohanZhangCMU

File: streaming/base/dataset.py [hunk 3]
```diff
@@ -526,7 +526,7 @@ def __init__(self,
         ]
         self._shm_prefix_int, self._locals_shm = get_shm_prefix(streams_local, streams_remote,
                                                                 self._unique_rank_world)
-        self._filelock_root = os.path.join(gettempdir(), 'streaming')
```
Comment (snarayan21) [inline streaming/base/dataset.py:529]: why this change?

File: streaming/base/shared/prefix.py [hunk 4]
```diff
@@ -108,8 +110,22 @@ def _check_and_find(streams_local: list[str], streams_remote: list[Union[str, No
...
         name = _get_path(prefix_int, LOCALS)
+
+        # Check if any shared memory filelocks exist for the current prefix
+        try:
+            shared_memory_exists = any(
+                os.path.exists(os.path.join(gettempdir(), _get_path(prefix_int, shm_name)))
+                for shm_name in SHM_TO_CLEAN)
+            if shared_memory_exists:
+                continue
+        except PermissionError:
```
Comment (snarayan21) [inline streaming/base/shared/prefix.py:121]: at which point would the permission error be thrown? when we check if the path exists?

**Review by snarayan21 — COMMENTED** [review]
seems fine to me...i think. would like to see testing

File: tests/test_shared.py [hunk 5]
```diff
@@ -157,3 +161,18 @@ def test_shared_array_size_is_integer(mock_shared_memory: MagicMock, dtype: type
     mock_shared_memory.assert_called_once()  # pyright: ignore
     size_arg = mock_shared_memory.call_args[1]['size']
     assert isinstance(size_arg, int), 'Size passed to SharedMemory is not an integer'
+
+
+def test_check_and_find_skips_filelock_conflict():
```
Comment (karan6181) [inline tests/test_shared.py:166]: Can we also add a mock test for PermissionError?

File: streaming/base/shared/prefix.py [hunk 6]
```diff
@@ -127,15 +155,16 @@ def _check_and_find(streams_local: list[str], streams_remote: list[Union[str, No
                         raise ValueError(
                             f'Reused local directory: {streams_local} vs ' +
                             f'{their_locals}. Provide a different one. If using ' +
-                            f'a unique local directory, try deleting the local directory and ' +
-                            f'call `streaming.base.util.clean_stale_shared_memory()` only once ' +
-                            f'in your script to clean up the stale shared memory before ' +
+                            f'a unique local directory, try deleting the local directory and '
```
Comment (karan6181) [inline streaming/base/shared/prefix.py:158]: Please revert this change, the `+` went to the next line which is not required

**Comment by XiaohanZhangCMU** [comment]
@karan6181 @snarayan21 @bigning testings are done. PR is ready for review then merge. PTAL.

File: streaming/base/shared/prefix.py (hunk 6 above)
Comment (snarayan21) [reply streaming/base/shared/prefix.py]: @XiaohanZhangCMU is this resolved / an issue?

File: streaming/base/shared/prefix.py [hunk 7]
```diff
@@ -184,29 +210,37 @@ def get_shm_prefix(streams_local: list[str],
     # Check my locals for overlap.
     _check_self(streams_local)
 
+    prefix_int = max([
```
Comment (snarayan21) [inline streaming/base/shared/prefix.py:213]: is there a case where the found prefix ints would be different for the different shm names? is this the duplicate shmem stuff / not getting cleaned thing you mention in the description?

**Review by snarayan21 — COMMENTED** [review]
couple questions, but seems fine to me overall. Deferring to @karan6181 and @bigning on approval

File: streaming/base/shared/prefix.py (hunk 7 above)
Comment (karan6181) [reply streaming/base/shared/prefix.py:213]: > is there a case where the found prefix ints would be different for the different shm names?

I strongly doubt that this will ever happen.

File: streaming/base/shared/prefix.py [hunk 8]
```diff
@@ -184,29 +210,37 @@ def get_shm_prefix(streams_local: list[str],
...
+        dist.barrier()
+
     # First, the local leader registers the first available shm prefix, recording its locals.
     if world.is_local_leader:
-        prefix_int = _check_and_find_retrying(streams_local, streams_remote, retry)
         name = _get_path(prefix_int, LOCALS)
         data = _pack_locals(streams_local, prefix_int)
         shm = SharedMemory(name, True, len(data))
         shm.buf[:len(data)] = data
+        their_locals, their_prefix_int = _unpack_locals(bytes(shm.buf))
```
Comment (karan6181) [inline streaming/base/shared/prefix.py:227]: We do not need this change right?

File: streaming/base/shared/prefix.py [hunk 9]
```diff
@@ -184,29 +210,37 @@ def get_shm_prefix(streams_local: list[str],
...
+        name = _get_path(prefix_int, LOCALS)
+        try:
+            shm = SharedMemory(name, False)
+        except FileNotFoundError:
+            raise RuntimeError(f'Internal error: shared memory prefix was not registered by ' +
+                               f'local leader. This may be because you specified ' +
+                               f'different ``local`` parameters from different ranks.')
+
+        their_locals, their_prefix_int = _unpack_locals(bytes(shm.buf))
+        if streams_local != their_locals or prefix_int != their_prefix_int:
```
Comment (karan6181) [inline streaming/base/shared/prefix.py:243]: Should we add a debug log to print all these variables before we raise an exception?

File: streaming/base/shared/prefix.py [hunk 10]
```diff
@@ -184,29 +210,37 @@ def get_shm_prefix(streams_local: list[str],
     # Check my locals for overlap.
     _check_self(streams_local)
 
+    prefix_int = max([
+        _check_and_find_retrying(streams_local, streams_remote, shm_name=shm_name, retry=retry)
```
Comment (bigning) [inline streaming/base/shared/prefix.py:214]: Just for my learning, why do we need retry here? is there any exception that not captured inside `_check_and_find`?

File: streaming/base/shared/prefix.py (hunk 9 above)
Comment (bigning) [reply streaming/base/shared/prefix.py:243]: or maybe print these variables in the RuntimeError message.

**Review by bigning — COMMENTED** [review]
thank you for the fix! Will let @karan6181  stamp!

File: streaming/base/shared/prefix.py (hunk 9 above)
Comment (XiaohanZhangCMU) [reply streaming/base/shared/prefix.py:242]: Added to err msg.

File: streaming/base/shared/prefix.py (hunk 10 above)
Comment (XiaohanZhangCMU) [reply streaming/base/shared/prefix.py:214]: @karan6181 may know better why it was there. 
My understanding is that there can be some racing when multiple processes try to attach the same shm file at the same time.

Comment (karan6181) [reply streaming/base/shared/prefix.py:214]: There may be some old processes attempting to clean up while the new process tries to attach to shared memory. The retries allow the old processes to finish their tasks.

**Review by karan6181 — APPROVED** [review]
LGTM. Thank You!
